scanners/scannerabc.py

```diff
@@ -0,0 +1,43 @@
+from abc import ABCMeta, abstractmethod
+from typing import Type, TypeVar
+import os
+
+ScannerABCT = TypeVar("ScannerABCT")
+
+
+class ScannerABC(metaclass=ABCMeta):
+
+    def __init__(self, domain: str, handles: dict, environment: dict,
+                 options: dict, extra: dict={}) -> None:
+        self.domain = domain
+        self.handles = handles
+        self.environment = {**environment, **extra}
+        self.options = options
+        self.report_dir = self.options.get("output", "./")
+        self.cache_dir = os.path.join(self.report_dir, "cache")
+        self.results_dir = os.path.join(self.report_dir, "results")
+
+        # Some subclasses may define an ``extra_environment`` property in their
+        # __init__ methods, and we want to fold its values into the
+        # ``environment`` property.
+        self.environment.update(getattr(self, "extra_environment", {}))
```

Comment:
There hasn't been an `extra` or `extra_environment` value used in scanners. I'd prefer to keep the number of moving parts smaller here if possible.

Reply:
The current structure of `scan` effectively passes around global variables that act as an extra environment. It's possible there's an elegant way to refactor it, but I've tried to change the structure of `scan` only minimally in this first pass. I'll take another pass through it and try to simplify this.